### PR TITLE
Week #4: Authentication -- Add ability for user to sign in and out.

### PR DESCRIPTION
**Pull request by JabariD**
Authentication is a key topic in many applications, so I included it in my portfolio starter project.

This adds the ability for the user to sign in and out. We also have a welcome message that displays depending if the user is signed in or not.

WIP: 1) User can pick a username 2) User can write their own todos. 3) User can display others user's todos.
Ambitions: 1) Users can specify if guest can see their todos 
 
Link: https://paytondennis-step-2020.uc.r.appspot.com/index.html
@ptai7 @chrisc94

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/auth/LogoutServlet.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/* Additons */
+import com.google.gson.Gson;
+import java.util.ArrayList;
+
+/** Log the user in. Redirect back to the main page! */
```
Comment (ptai7) [inline portfolio/src/main/java/com/google/sps/auth/LogoutServlet.java:29]: Update this server to reflect this LogoutServlet's purpose.

File: portfolio/src/main/java/com/google/sps/auth/LogoutServlet.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+@WebServlet("/logout")
+public class LogoutServlet extends HttpServlet {
+    private ArrayList<String> data = new ArrayList<String>();
+
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+            UserService userService = UserServiceFactory.getUserService();
+            if (userService.isUserLoggedIn()) {
```
Comment (ptai7) [inline portfolio/src/main/java/com/google/sps/auth/LogoutServlet.java:37]: De-indent this line? for better formatting.

File: portfolio/src/main/java/com/google/sps/todos/StoreQuantityServlet.java [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+import java.io.FileReader;
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.Reader;
+import com.google.gson.GsonBuilder;
+import java.util.ArrayList;
+/* -------- */
+
+/* Datastore */
+// import com.google.appengine.api.datastore.DatastoreService;
```
Comment (ptai7) [inline portfolio/src/main/java/com/google/sps/todos/StoreQuantityServlet.java:37]: Remove unused comments.

File: portfolio/src/main/java/com/google/sps/todos/StoreQuantityServlet.java [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
+        try {
+            // Retrieve from Datastore if avaliable.
+            Query query = new Query("Quantity");
+            PreparedQuery results = datastore.prepare(query);
+            for (Entity entity : results.asIterable())
+                numberOfTodos = (Integer) entity.getProperty("num");
+        } catch (Exception e) {
+
```
Comment (ptai7) [inline portfolio/src/main/java/com/google/sps/todos/StoreQuantityServlet.java:70]: Add console printing to indicate that retrieval failed.

File: portfolio/src/main/java/com/google/sps/todos/StoreQuantityServlet.java [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+        StringBuilder buffer = new StringBuilder();
+        BufferedReader reader = request.getReader();
+        String line;
+        while ((line = reader.readLine()) != null) {
+            buffer.append(line);
+        }
+        String data = buffer.toString();
+       
+        try {
+            numberOfTodos = Integer.parseInt(data.substring(13,14));
```
Comment (ptai7) [inline portfolio/src/main/java/com/google/sps/todos/StoreQuantityServlet.java:95]: Why is this hardcoded to (13,14)? Either add a comment or use more general string parsing.

File: portfolio/src/main/webapp/login.js [hunk 6]
```diff
@@ -0,0 +1,31 @@
+
+/** Once we click the 'Sign In' button, we sign the user in. */
+async function signIn() {
+    // Determine if user is already logged in.
+    if (document.getElementById("email").innerHTML.localeCompare("Welcome, Guest.") === 0) {
```
Comment (ptai7) [inline portfolio/src/main/webapp/login.js:5]: Is there a way where you can reuse checkIfSignedIn() instead of checking against the HTML?

**Comment by JabariD** [comment]
Link: https://paytondennis-step-2020.uc.r.appspot.com/index.html

File: portfolio/src/main/java/com/google/sps/auth/UserServlet.java [hunk 7]
```diff
@@ -26,23 +26,45 @@
 import com.google.gson.Gson;
 import java.util.ArrayList;
 /* -------- */
+import com.google.appengine.api.datastore.*;
 
 /** Simply check if the user is signed in and return the email. */
 @WebServlet("/user")
 public class UserServlet extends HttpServlet {
     private ArrayList<String> data = new ArrayList<String>();
+    private Boolean x = false;
```
Comment (ptai7) [inline portfolio/src/main/java/com/google/sps/auth/UserServlet.java:35]: What is "x" and where is it used?

Comment (JabariD) [reply portfolio/src/main/java/com/google/sps/auth/UserServlet.java:35]: So sorry! I forgot to delete that, it was for testing purposes 👍

**Pull request merged**